metadata/src/main/java/org/apache/kafka/image/loader/MetadataBatchLoader.java

```diff
@@ -0,0 +1,257 @@
+                    batch.baseOffset() + indexWithinBatch, e);
+            }
+
+            // Emit the accumulated delta if a new transaction has been started and one of the following is true
+            // 1) this is not the first record in this batch
+            // 2) this is not the first batch since last emitting a delta
+            if (transactionState == TransactionState.STARTED_TRANSACTION && (indexWithinBatch > 0 || numBatches > 0)) {
+                MetadataProvenance provenance = new MetadataProvenance(lastOffset, lastEpoch, lastContainedLogTimeMs);
+                LogDeltaManifest manifest = LogDeltaManifest.newBuilder()
+                    .provenance(provenance)
+                    .leaderAndEpoch(leaderAndEpochSupplier.get())
+                    .numBatches(numBatches) // This will be zero if we have not yet read a batch
+                    .elapsedNs(totalBatchElapsedNs)
+                    .numBytes(numBytes)     // This will be zero if we have not yet read a batch
+                    .build();
+                if (log.isDebugEnabled()) {
+                    log.debug("handleCommit: Generated a metadata delta between {} and {} from {} batch(es) in {} us.",
+                            image.offset(), manifest.provenance().lastContainedOffset(),
+                            manifest.numBatches(), NANOSECONDS.toMicros(manifest.elapsedNs()));
+                }
+                applyDeltaAndUpdate(delta, manifest);
+                transactionState = TransactionState.STARTED_TRANSACTION;
+            }
+
+            lastOffset = batch.baseOffset() + indexWithinBatch;
+            indexWithinBatch++;
+        }
+        long elapsedNs = time.nanoseconds() - startNs;
+
+        // Update state for the manifest. The actual byte count will only be included in the last delta emitted for
+        // a given batch or transaction.
+        lastOffset = batch.lastOffset();
+        numBytes += batch.sizeInBytes();
+        numBatches += 1;
+        totalBatchElapsedNs += elapsedNs;
+        return totalBatchElapsedNs;
+    }
+
+    /**
+     * Flush the metadata accumulated in this batch loader if not in the middle of a transaction. The
+     * flushed metadata will be passed to the {@link MetadataUpdater} configured for this class.
+     */
+    public void maybeFlushBatches() {
+        MetadataProvenance provenance = new MetadataProvenance(lastOffset, lastEpoch, lastContainedLogTimeMs);
+        LogDeltaManifest manifest = LogDeltaManifest.newBuilder()
+            .provenance(provenance)
+            .leaderAndEpoch(leaderAndEpochSupplier.get())
+            .numBatches(numBatches)
+            .elapsedNs(totalBatchElapsedNs)
+            .numBytes(numBytes)
+            .build();
+        switch (transactionState) {
+            case STARTED_TRANSACTION:
+            case CONTINUED_TRANSACTION:
+                log.debug("handleCommit: not publishing since a transaction is still in progress");
```

Comment:
I think we should have some basic info in the log messages where we're not publishing. Basically, like the stuff we have in the handleCommit message.

"A delta between XXXX and YYYY, from ZZZZ batches"